src/Runner.Worker/Handlers/CompositeActionHandler.cs

```diff
@@ -153,29 +143,29 @@ private async Task RunStepsAsync(List<IStep> compositeSteps)
         {
             ArgUtil.NotNull(compositeSteps, nameof(compositeSteps));
 
-            // The parent StepsRunner of the whole Composite Action Step handles the cancellation stuff already. 
             foreach (IStep step in compositeSteps)
             {
-                Trace.Info($"Processing composite step: DisplayName='{step.DisplayName}'");
+                Trace.Info($"Processing nested step: DisplayName='{step.DisplayName}'");
 
                 step.ExecutionContext.ExpressionValues["steps"] = ExecutionContext.Global.StepsContext.GetScope(step.ExecutionContext.ScopeName);
 
-                // Populate env context for each step
-                Trace.Info("Initialize Env context for step");
+                // Initialize env context
+                Trace.Info("Initialize Env context for nested step");
 #if OS_WINDOWS
                 var envContext = new DictionaryContextData();
 #else
                 var envContext = new CaseSensitiveDictionaryContextData();
 #endif
+                step.ExecutionContext.ExpressionValues["env"] = envContext;
 
-                // Global env
+                // Merge global env
                 foreach (var pair in ExecutionContext.Global.EnvironmentVariables)
                 {
                     envContext[pair.Key] = new StringContextData(pair.Value ?? string.Empty);
                 }
 
-                // Stomps over with outside step env
-                if (step.ExecutionContext.ExpressionValues.TryGetValue("env", out var envContextData))
+                // Merge composite-step env
+                if (ExecutionContext.ExpressionValues.TryGetValue("env", out var envContextData))
```

Comment:
Simplified this a little bit to remove the indirection (i.e. previously `ExecutionContext.CreateCompositeStep` copied the composite step's env to the child step)